tool/tsh/db.go

```diff
@@ -124,33 +124,42 @@ func databaseLogin(cf *CLIConf, tc *client.TeleportClient, db tlsca.RouteToDatab
 		// ref: https://redis.io/commands/auth
 		db.Username = defaults.DefaultRedisUsername
 	}
-	profile, err := client.StatusCurrent(cf.HomePath, cf.Proxy)
+
+	// Note: we don't use StatusCurrentWithIdentity here as virtual identity
+	// profiles are read-only.
+	profile, err := client.StatusCurrentWithIdentity(cf.HomePath, cf.Proxy, cf.IdentityFileIn)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	var key *client.Key
-	if err = client.RetryWithRelogin(cf.Context, tc, func() error {
-		key, err = tc.IssueUserCertsWithMFA(cf.Context, client.ReissueParams{
-			RouteToCluster: tc.SiteName,
-			RouteToDatabase: proto.RouteToDatabase{
-				ServiceName: db.ServiceName,
-				Protocol:    db.Protocol,
-				Username:    db.Username,
-				Database:    db.Database,
-			},
-			AccessRequests: profile.ActiveRequests.AccessRequests,
-		})
-		return trace.Wrap(err)
-	}); err != nil {
-		return trace.Wrap(err)
-	}
-	if err = tc.LocalAgent().AddDatabaseKey(key); err != nil {
-		return trace.Wrap(err)
+	// Identity files contain themselves act as the database credentials (if
+	// any), so don't bother fetching new certs.
+	if profile.IsVirtual {
+		log.Info("Note: already logged in due to an identity file (`-i ...`); will only update database config files.")
+	} else {
+		var key *client.Key
+		if err = client.RetryWithRelogin(cf.Context, tc, func() error {
```

Comment:
Should we perhaps add virtual (read-only?) profile handling to `RetryWithLogin` to have uniform support across the various ops?

Reply:
I'd like to give this a deeper look. My gut feeling is that it'd be similarly verbose: everything that issues certs would need to see the identity file, skip reissuing certs, and then store to memory rather than disk. We'd still have issues with profiles and especially the paths that get passed to e.g. database CLIs that are expected to be on disk.

Reply:
I've looked a bit more, and I don't think we really save anything here: `RetryWithRelogin()` (and more specifically, `client.Login()`) just manages the core client identity which we already configure with the identity file at startup.

Database / Kubernetes / App access all use profile certs and never try to use the core identity even if it does actually have the proper permissions. Unfortunately the only shared codepath for these is the profiles system, which is what necessitated this hacky solution to begin with.

Ideally I'd like to see us invest in a larger refactoring to clean all of this up but frankly I don't even know what I'd want it to look like, tsh has a lot of functionality to juggle. In the mean time hopefully we can enable at least the Machine ID use-case.